src/main/java/org/traccar/WebDataHandler.java

```diff
@@ -155,32 +182,92 @@ public String formatRequest(Position position) throws UnsupportedEncodingExcepti
+                        requestBuilder.header(values[0].trim(), values[1].trim());
+                    }
+                }
+
+                if (json) {
+                    payload = prepareJsonPayload(position);
+                }
+
+                deliveryPending.incrementAndGet();
+
+                send();
             }
-        }
 
-        Invocation.Builder requestBuilder = client.target(url).request();
+            private void send() {
+                if (json) {
+                    requestBuilder.async().post(Entity.json(payload), this);
+                } else {
+                    requestBuilder.async().get(this);
+                }
+            }
 
-        if (header != null && !header.isEmpty()) {
-            for (String line: header.split("\\r?\\n")) {
-                String[] values = line.split(":", 2);
-                requestBuilder.header(values[0].trim(), values[1].trim());
+            private void retry() {
+                boolean ok = false;
+                try {
+                    if (retryEnabled && deliveryPending.get() <= retryLimit && retries < retryCount) {
+                        Main.getInjector().getInstance(Timer.class).newTimeout(new TimerTask() {
+                            @Override
+                            public void run(Timeout timeout) {
+                                boolean ok = false;
+                                try {
+                                    if (!timeout.isCancelled()) {
+                                        send();
+                                        ok = true;
+                                    }
+                                } finally {
+                                    if (!ok) {
+                                        deliveryPending.decrementAndGet();
+                                    }
+                                }
+                            }
+                        }, retryDelay * (int) Math.pow(2, retries++), TimeUnit.MILLISECONDS);
+                        ok = true;
+                    }
+                } finally {
+                    int pending = ok ? deliveryPending.get() : deliveryPending.decrementAndGet();
+                    LOGGER.warn("Position forwarding failed: " + pending + " pending");
+                }
+            }
+
+            @Override
+            public void completed(Response response) {
+                if (response.getStatusInfo().getFamily() == Response.Status.Family.SUCCESSFUL) {
+                    deliveryPending.decrementAndGet();
+                } else {
+                    retry();
+                }
+            }
+
+            @Override
+            public void failed(Throwable throwable) {
+                retry();
             }
-        }
 
-        if (json) {
-            requestBuilder.async().post(Entity.json(prepareJsonPayload(position)));
-        } else {
-            requestBuilder.async().get();
         }
 
+        new AsyncRequestAndCallback(position);
```

Comment:
It seems weird that constructor calls `send` on its own. I think you should probably call send here instead.

Reply:
Ok. Let's do it that way.